# PD01 · Explorar y preparar AGROSAVIA

[Índice y distribución de las cuatro horas](../ejercicios/README.md) · [Guía de pandas](../docs/manual-pandas.md)

**Tiempo de trabajo: 60 minutos.** El objetivo es reconocer la estructura de un DataFrame y preparar medidas y fechas conservando los valores originales.

| Símbolo | Uso |
|:---:|---|
| 📘 | Concepto o función de apoyo. |
| ✅ | Evidencia que debe quedar visible. |
| ⚠️ | Límite de interpretación. |

Esta plantilla se encuentra en `ejercicios/PD01_exploracion_agrosavia.ipynb`. Antes de comenzar, se crea una copia en `soluciones/PD01_exploracion_agrosavia.ipynb` dentro del repositorio de cada estudiante. El desarrollo y la entrega se realizan en esa copia; la plantilla de `ejercicios/` se conserva como enunciado. En JupyterLab se puede duplicar el archivo y mover la copia a `soluciones/`, manteniendo el nombre indicado; si ya existe una solución, se continúa sobre ella. Las instrucciones se presentan en celdas Markdown; las celdas de código marcadas como **Desarrollo** contienen el espacio de trabajo. Cada actividad incluye una celda para explicar el resultado.

Se utiliza el kernel del entorno `.venv` del curso. Los CSV ya deben estar descargados mediante `kit/00_datos.py`. La celda inicial prepara las rutas y carga únicamente las entradas de este ejercicio; el desarrollo del análisis corresponde al estudiante. Este notebook se ejecuta de forma independiente.

**Apoyo en el manual:** secciones 1–5 y la receta de registros en revisión de la sección 11.

## Ubicación de trabajo: la raíz del repositorio

📘 **La raíz del repositorio** es la carpeta principal de la copia del proyecto. En la instalación del curso se llama `big-data-postgraduate` y contiene `README.md`, `kit/`, `ejercicios/` y `soluciones/`:

```text
bigdata/
└── big-data-postgraduate/       ← raíz del repositorio
    ├── README.md
    ├── kit/
    ├── ejercicios/              ← enunciados y plantillas
    └── soluciones/              ← notebooks resueltos
```

**Desde la terminal de Ubuntu en WSL.** `pwd` muestra la carpeta actual, `cd` cambia de carpeta y `ls` muestra su contenido. Los siguientes comandos se ejecutan en la terminal, no en una celda Python:

```bash
pwd
cd /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate
pwd
ls
```

✅ Después de `cd`, el segundo `pwd` debe mostrar `/mnt/c/Users/TUPTC/bigdata/big-data-postgraduate`. En la salida de `ls` deben aparecer `README.md`, `kit`, `ejercicios` y `soluciones`. La carpeta `bigdata` es la carpeta contenedora; la raíz del proyecto es su subcarpeta `big-data-postgraduate`.

La ruta anterior corresponde a la configuración de la guía WSL. Si el repositorio se clonó en otra ubicación, se sustituye por la ruta real de esa copia. Si `cd` informa que la carpeta no existe, se revisa esa ubicación antes de continuar.

**Desde una subcarpeta conocida.** `..` significa «carpeta superior». Se comprueba primero la ubicación con `pwd` y se utiliza la alternativa correspondiente:

| Ubicación actual, relativa al repositorio | Comando para volver a la raíz |
|---|---|
| `ejercicios/`, `soluciones/` o `kit/` | `cd ..` |
| `kit/Notebooks/` | `cd ../..` |

Al llegar a la raíz, se puede activar el entorno y abrir JupyterLab desde la terminal:

```bash
source .venv/bin/activate
jupyter lab
```

**Dentro de un notebook ya abierto.** La terminal y el kernel de Python mantienen sus propios directorios de trabajo. La celda de carga siguiente localiza la raíz y la guarda en `REPO`, aunque el notebook esté abierto desde `soluciones/` o `ejercicios/`. Después de ejecutar esa celda, el siguiente bloque puede ejecutarse en una celda Python para comprobar la ubicación y cambiar el directorio del kernel a la raíz:

```python
import os
from pathlib import Path

print("Directorio antes del cambio:", Path.cwd())
print("Raíz detectada:", REPO)
os.chdir(REPO)
print("Directorio después del cambio:", Path.cwd())
```

✅ `Path.cwd()` debe coincidir con `REPO` después del cambio. Esta operación cambia la base para interpretar rutas relativas; el archivo del notebook permanece en `soluciones/`. La carga de datos de estos ejercicios también funciona sin ese cambio porque construye las rutas a partir de `REPO`.

## Preparación

Se ejecuta la siguiente celda y se comprueba que el intérprete pertenece al entorno del curso y que pandas corresponde a la versión 2.2.3. Las rutas se calculan desde el repositorio, sin introducir rutas personales de Windows, WSL o macOS.

In [2]:
from pathlib import Path
import sys
import pandas as pd
from IPython.display import display

actual = Path.cwd().resolve()
REPO = next(
    (ruta for ruta in [actual, *actual.parents]
     if (ruta / "kit/fuentes.json").is_file()
     and (ruta / "kit/00_datos.py").is_file()),
    None,
)
if REPO is None:
    raise FileNotFoundError("El notebook debe abrirse dentro del repositorio del curso")

RAW = REPO / "kit/data/raw"
OUT = REPO / "kit/salidas/soluciones/PD01"
archivos = ['agrosavia.csv']
faltantes = [nombre for nombre in archivos if not (RAW / nombre).is_file()]
if faltantes:
    raise FileNotFoundError(f"Faltan CSV en {RAW}: {faltantes}")
OUT.mkdir(parents=True, exist_ok=True)

opciones = dict(sep=",", encoding="utf-8-sig", dtype=str, keep_default_na=False)
suelo = pd.read_csv(RAW / "agrosavia.csv", **opciones)


print("Python:", sys.version.split()[0])
print("Intérprete:", sys.executable)
print("pandas:", pd.__version__)
print("Repositorio:", REPO)
print("Salidas:", OUT)

Python: 3.12.15
Intérprete: /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate/.venv/bin/python
pandas: 2.2.3
Repositorio: /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate
Salidas: /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate/kit/salidas/soluciones/PD01


## 1. Inspección inicial · 10 minutos

Presentar la cantidad de filas y columnas, los nombres de las columnas, sus tipos y cinco registros. Explicar qué representa una fila y por qué la lectura inicial conserva los campos como texto.

📘 **Apoyo:** `shape`, `columns`, `dtypes`, `head` e `info`.

✅ **Comprobación:** El corte del manifiesto contiene 92.738 filas y 32 columnas. La tabla original debe conservar esa cantidad de filas durante el ejercicio.

In [8]:
# Desarrollo de la actividad 1

# Mostrar dimensiones, tipos y primeros registros de suelo.
print("Filas y columnas:", suelo.shape)
print("Filas:", suelo.shape[0])
print("Columnas:", suelo.shape[1])
print("=====================================================")
# Guardar el total de filas original para los controles posteriores.
filas_originales = suelo.shape[0]
print("Total de filas original:", filas_originales)
print("=====================================================")
#Mostrar columnas del df
print(list(suelo.columns))
print("=====================================================")
#Mostrar tipos del df
print(suelo.dtypes)
print("=====================================================")
#5 primeros registros
display(suelo.head())
print("=====================================================")
#Resumen general
suelo.info()
print("=====================================================")
#Comprobación
assert suelo.shape == (92738, 32), f"Se obtuvo {suelo.shape}"
print("92.738 filas y 32 columnas")

Filas y columnas: (92738, 32)
Filas: 92738
Columnas: 32
Total de filas original: 92738
['Secuencial', 'Fecha de Análisis', 'Departamento', 'Municipio', 'Cultivo', 'Estado', 'Tiempo de establecimiento', 'Topografia', 'Drenaje', 'Riego', 'Fertilizantes aplicados', 'pH agua:suelo', 'Materia organica', 'Fósforo Bray II', 'Azufre Fosfato monocalcico', 'Acidez Intercambiable', 'Aluminio intercambiable', 'Calcio intercambiable', 'Magnesio intercambiable', 'Potasio intercambiable', 'Sodio intercambiable', 'capacidad de intercambio cationico', 'Conductividad electrica', 'Hierro disponible olsen', 'Cobre disponible', 'Manganeso disponible Olsen', 'Zinc disponible Olsen', 'Boro disponible', 'Hierro disponible doble acido', 'Cobre disponible doble acido', 'Manganeso disponible doble acido', 'Zinc disponible doble \xa0acido']
Secuencial                            object
Fecha de Análisis                     object
Departamento                          object
Municipio                             ob

,Secuencial,Fecha de Análisis,Departamento,Municipio,Cultivo,Estado,Tiempo de establecimiento,Topografia,Drenaje,Riego,...,Conductividad electrica,Hierro disponible olsen,Cobre disponible,Manganeso disponible Olsen,Zinc disponible Olsen,Boro disponible,Hierro disponible doble acido,Cobre disponible doble acido,Manganeso disponible doble acido,Zinc disponible doble acido
0,1,7/01/2014,NARIÑO,SAN ANDRÉS DE TUMACO,No Indica,No indica,No indica,No indica,No indica,No indica,...,0.133,66.39,1.700,1.700,1.5,0.187,ND,ND,ND,ND
1,2,21/08/2014,HUILA,SANTA MARÍA,Granadilla,Establecido,De 1 a 5 años,Pendiente,Buen drenaje,No Tiene,...,2.749,250.9,3.7,29.4,28.70,2.084,ND,ND,ND,ND
2,3,22/08/2014,ANTIOQUIA,LIBORINA,Café,Por establecer,No indica,Pendiente,Mal drenaje,No Tiene,...,0.328,390,11.10,8.299,5,0.085,ND,ND,ND,ND
3,4,22/08/2014,ANTIOQUIA,LIBORINA,Maracuyá,Por establecer,No indica,Ondulado,Mal drenaje,No Tiene,...,0.171,200,4.2,4.699,1,0.229,ND,ND,ND,ND
4,5,22/08/2014,ANTIOQUIA,LIBORINA,Café,Por establecer,No indica,Ondulado,Mal drenaje,No Tiene,...,0.323,117.9,3.8,4.600,0.700,0.139,ND,ND,ND,ND


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92738 entries, 0 to 92737
Data columns (total 32 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Secuencial                          92738 non-null  object
 1   Fecha de Análisis                   92738 non-null  object
 2   Departamento                        92738 non-null  object
 3   Municipio                           92738 non-null  object
 4   Cultivo                             92738 non-null  object
 5   Estado                              92738 non-null  object
 6   Tiempo de establecimiento           92738 non-null  object
 7   Topografia                          92738 non-null  object
 8   Drenaje                             92738 non-null  object
 9   Riego                               92738 non-null  object
 10  Fertilizantes aplicados             92738 non-null  object
 11  pH agua:suelo                       92738 non-null  ob

**Interpretación del resultado:**

La tabla `suelo` tiene **92.738 filas y 32 columnas**, que coincide con el corte del manifiesto. Cada fila es un análisis de suelo individual identificado por `Secuencial` con su fecha, ubicación, cultivo, manejo y resultados de laboratorio. Esa cantidad queda guardada en `filas_originales` para verificar que ninguna limpieza posterior elimine registros.

Las 32 columnas aparecen como `object` porque la lectura usó `dtype=str` y `keep_default_na=False`, se conservan los valores tal como vienen en el archivo, sin conversiones automáticas, y los vacíos quedan como `""` en lugar de `NaN`. Por eso `info()` muestra 92.738 valores "non-null" en todas las columnas; los datos faltantes reales todavía no son visibles.

Además, se observa que algunos nombres de columna necesitarán limpieza: por ejemplo, `Zinc disponible doble  acido` contiene un espacio no separable (`\xa0`) y doble espacio, y hay mezcla de mayúsculas y tildes en otros nombres.

## 2. Diagnóstico de cinco campos · 10 minutos

Construir una tabla llamada `perfil` para `Secuencial`, `Fecha de Análisis`, `Departamento`, `Municipio` y `pH agua:suelo`. Para cada campo se registran los nulos de pandas, los textos vacíos después de retirar espacios y los marcadores `ND`. Las tres situaciones se cuentan por separado.

📘 **Apoyo:** `isna`, `.str.strip`, `.str.upper`, `.eq` y `sum`.

✅ **Comprobación:** `perfil` debe tener una fila por campo. Con `keep_default_na=False`, un texto vacío o `ND` no se convierte automáticamente en un nulo.

In [9]:
# Desarrollo de la actividad 2
campos = ['Secuencial', 'Fecha de Análisis', 'Departamento', 'Municipio', 'pH agua:suelo']
print(campos)
print("=====================================================")

#Se suman los nulos
nulos = suelo[campos].isna().sum()
print(nulos)
print("=====================================================")

#Se aplica ciclo para contar vacios
vacios = suelo[campos].apply(lambda col: col.str.strip().eq("")).sum()
print(vacios)
print("=====================================================")

#Se cuentan nd haciendo transformaciones de texto para matchear
nd = suelo[campos].apply(lambda col: col.str.strip().str.upper().eq("ND")).sum()
print(nd)
print("=====================================================")

# Crear y mostrar perfil con las cinco columnas solicitadas.
perfil = pd.DataFrame({
    "campo": campos,
    "nulos_pandas": nulos.values,
    "vacios_tras_strip": vacios.values,
    "marcador_ND": nd.values,
})
display(perfil)
print("=====================================================")

#comprobación
assert len(perfil) == 5, f"Se esperaban 5 filas y hay {len(perfil)}"
print("Perfil tiene una fila por campo")

['Secuencial', 'Fecha de Análisis', 'Departamento', 'Municipio', 'pH agua:suelo']
Secuencial           0
Fecha de Análisis    0
Departamento         0
Municipio            0
pH agua:suelo        0
dtype: int64
Secuencial           0
Fecha de Análisis    0
Departamento         0
Municipio            2
pH agua:suelo        0
dtype: int64
Secuencial           0
Fecha de Análisis    0
Departamento         0
Municipio            0
pH agua:suelo        0
dtype: int64


,campo,nulos_pandas,vacios_tras_strip,marcador_ND
0,Secuencial,0,0,0
1,Fecha de Análisis,0,0,0
2,Departamento,0,0,0
3,Municipio,0,2,0
4,pH agua:suelo,0,0,0


Perfil tiene una fila por campo


**Interpretación del resultado:**

La tabla `perfil` resume tres tipos de dato faltante en cinco campos. Ningún campo tiene nulos de pandas, lo cual es consistente con la lectura usando `keep_default_na=False`: los vacíos y los marcadores no se convierten automáticamente en `NaN`, por lo que `isna()` por sí solo no detecta nada.

Al contar los textos vacíos después de retirar espacios, aparece un único hallazgo: `Municipio` tiene **2 registros vacíos**. `Secuencial`, `Fecha de Análisis`, `Departamento` y `pH agua:suelo` no tienen vacíos, y ninguno de los cinco campos contiene el marcador `ND`.

## 
3. Conversión de pH y fecha · 15 minutos

Crear `ph` desde `pH agua:suelo` y `fecha_dt` desde `Fecha de Análisis`. Utilizar conversión numérica con `errors="coerce"` y fecha con formato `%d/%m/%Y`. Conservar los originales. Distinguir entradas vacías de textos no vacíos que no pudieron convertirse.

📘 **Apoyo:** `pd.to_numeric`, `.str.replace`, `pd.to_datetime`, `isna` y máscaras booleanas.

✅ **Comprobación:** Se muestran los conteos de conversiones fallidas y hasta cinco ejemplos con su valor original. No se reemplazan medidas ausentes por cero.

In [10]:
# Desarrollo de la actividad 3
# Crear suelo["ph"] y suelo["fecha_dt"].
ph_texto = suelo["pH agua:suelo"].str.strip().str.replace(",", ".", regex=False)
suelo["ph"] = pd.to_numeric(ph_texto, errors="coerce")
print(suelo["ph"].describe())
print("=====================================================")

#Convertir fecha
fecha_texto = suelo["Fecha de Análisis"].str.strip()
suelo["fecha_dt"] = pd.to_datetime(fecha_texto, format="%d/%m/%Y", errors="coerce")
print(suelo["fecha_dt"].min(), suelo["fecha_dt"].max())
print("=====================================================")

#Vacias y fallidos
ph_vacio = suelo["pH agua:suelo"].str.strip().eq("")
ph_fallido = suelo["ph"].isna() & ~ph_vacio

fecha_vacia = suelo["Fecha de Análisis"].str.strip().eq("")
fecha_fallida = suelo["fecha_dt"].isna() & ~fecha_vacia
print("=====================================================")

#Conteos
print("pH vacíos:", ph_vacio.sum())
print("pH fallidos:", ph_fallido.sum())
print("Fechas vacías:", fecha_vacia.sum())
print("Fechas fallidas:", fecha_fallida.sum())
print("=====================================================")

#Fallidos
print("Ejemplos pH fallidos:")
display(suelo.loc[ph_fallido, ["Secuencial", "pH agua:suelo", "ph"]].head(5))

print("Ejemplos fechas fallidas:")
display(suelo.loc[fecha_fallida, ["Secuencial", "Fecha de Análisis", "fecha_dt"]].head(5))
print("=====================================================")

# Mostrar los fallos de conversión y sus valores originales.
assert len(suelo) == filas_originales
assert suelo["ph"].isna().sum() == ph_vacio.sum() + ph_fallido.sum()
assert suelo["fecha_dt"].isna().sum() == fecha_vacia.sum() + fecha_fallida.sum()
print("✅ Sin filas perdidas; cada NaN/NaT es vacío o fallido; no se rellenó nada con cero")

count    92727.000000
mean         5.708232
std          1.019957
min          3.190000
25%          4.960000
50%          5.470000
75%          6.260000
max         10.570000
Name: ph, dtype: float64
2014-01-07 00:00:00 2025-09-05 00:00:00
pH vacíos: 0
pH fallidos: 11
Fechas vacías: 0
Fechas fallidas: 0
Ejemplos pH fallidos:


,Secuencial,pH agua:suelo,ph
7282,7283,4..01,NaN
46528,46516,>10.00,NaN
46530,46518,>10.00,NaN
52903,52901,<3.80,NaN
56330,56329,<3.80,NaN


Ejemplos fechas fallidas:


,Secuencial,Fecha de Análisis,fecha_dt


✅ Sin filas perdidas; cada NaN/NaT es vacío o fallido; no se rellenó nada con cero


**Interpretación del resultado:**

Se crearon `ph` y `fecha_dt` conservando las columnas originales. La tabla mantiene sus 92.738 filas.

En `pH agua:suelo` no hay entradas vacías, pero **11 textos no vacíos no pudieron convertirse**. Los ejemplos muestran dos causas: un error de digitación (`4..01`) y valores censurados del laboratorio (`>10.00`, `<3.80`), que indican un límite de detección y no una medida exacta. Estos casos quedan como `NaN` en `ph`; no se reemplazaron por cero ni por otro valor, porque eso introduciría medidas falsas.

En `Fecha de Análisis` no hubo vacíos ni conversiones fallidas: las 92.738 fechas se interpretaron con el formato `%d/%m/%Y`

## 4. Registros para revisión · 15 minutos

Crear la columna booleana `en_revision`. Se marca una fila si su pH es ausente, está fuera de `[0, 14]` o su fecha no se pudo interpretar. Contar por separado cada motivo y el total de filas marcadas. Conservar todas las observaciones.

📘 **Apoyo:** `between`, `isna`, `|`, `~`, `.loc` y `value_counts`.

✅ **Comprobación:** Las filas marcadas y no marcadas deben sumar el total original. Una misma fila puede tener varios motivos; por eso la suma de los motivos no necesariamente coincide con el total marcado.

In [12]:
# Desarrollo de la actividad 4
# Crear en_revision y los conteos por motivo.
motivo_ph_ausente = suelo["ph"].isna()
print("pH ausente:", motivo_ph_ausente.sum())

motivo_ph_fuera = suelo["ph"].notna() & ~suelo["ph"].between(0, 14)
print("pH fuera de [0, 14]:", motivo_ph_fuera.sum())

motivo_fecha = suelo["fecha_dt"].isna()
print("Fecha no interpretada:", motivo_fecha.sum())

suelo["en_revision"] = motivo_ph_ausente | motivo_ph_fuera | motivo_fecha
print(suelo["en_revision"].value_counts())

resumen = pd.DataFrame({
    "motivo": ["pH ausente", "pH fuera de [0, 14]", "Fecha no interpretada", "Total filas marcadas"],
    "filas": [motivo_ph_ausente.sum(), motivo_ph_fuera.sum(), motivo_fecha.sum(), suelo["en_revision"].sum()],
})
display(resumen)

# Comprobar la conservación del total original con assert.
marcadas = suelo["en_revision"].sum()
no_marcadas = (~suelo["en_revision"]).sum()
assert len(suelo) == filas_originales
assert marcadas + no_marcadas == filas_originales
print("Marcadas + no marcadas =", marcadas + no_marcadas)

pH ausente: 11
pH fuera de [0, 14]: 0
Fecha no interpretada: 0
en_revision
False    92727
True        11
Name: count, dtype: int64


,motivo,filas
0,pH ausente,11
1,"pH fuera de [0, 14]",0
2,Fecha no interpretada,0
3,Total filas marcadas,11


Marcadas + no marcadas = 92738


**Interpretación del resultado:**

Se creó la columna booleana `en_revision`, que marca una fila si su pH es ausente, está fuera de [0, 14] o su fecha no se pudo interpretar. No se eliminó ninguna observación: las 11 filas marcadas y las 92.727 no marcadas suman las 92.738 originales.

## 5. Evidencias y conclusiones · 10 minutos

Construir `controles` con las columnas `control` y `cantidad`: total de entrada, pH sin número, pH fuera del intervalo, fecha no interpretable y total de filas en revisión. Exportar a `OUT / "controles_agrosavia.csv"` sin índice. Redactar dos hallazgos apoyados en los conteos.

📘 **Apoyo:** `pd.DataFrame`, `to_csv(index=False)` y las tablas de las actividades anteriores.

✅ **Comprobación:** El archivo se genera dentro de la ruta `OUT`. Cada hallazgo menciona un conteo y explica su significado o limitación.

In [13]:
# Desarrollo de la actividad 5
# Construir controles, mostrarlo y exportarlo a OUT.
controles = pd.DataFrame({
    "control": [
        "Total de entrada",
        "pH sin número",
        "pH fuera del intervalo",
        "Fecha no interpretable",
        "Total de filas en revisión",
    ],
    "cantidad": [
        filas_originales,
        int(motivo_ph_ausente.sum()),
        int(motivo_ph_fuera.sum()),
        int(motivo_fecha.sum()),
        int(suelo["en_revision"].sum()),
    ],
})
display(controles)
#=================================================

ruta_controles = OUT / "controles_agrosavia.csv"
controles.to_csv(ruta_controles, index=False)
print("Archivo generado:", ruta_controles)
#=================================================

assert ruta_controles.is_file()
assert ruta_controles.parent == OUT
print("Archivo dentro de OUT")
display(pd.read_csv(ruta_controles))
# Redactar los dos hallazgos en la celda Markdown siguiente.

,control,cantidad
0,Total de entrada,92738
1,pH sin número,11
2,pH fuera del intervalo,0
3,Fecha no interpretable,0
4,Total de filas en revisión,11


Archivo generado: /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate/kit/salidas/soluciones/PD01/controles_agrosavia.csv
Archivo dentro de OUT


,control,cantidad
0,Total de entrada,92738
1,pH sin número,11
2,pH fuera del intervalo,0
3,Fecha no interpretable,0
4,Total de filas en revisión,11


**Interpretación del resultado:**

La tabla `controles` resume en cinco conteos la calidad de la entrada y se exportó sin índice a `OUT / "controles_agrosavia.csv"`. De las **92.738** filas de entrada, **11** tienen un pH sin número, **0** tienen pH fuera de [0, 14], **0** tienen fecha no interpretable y **11** quedan en revisión. Ninguna fila se eliminó.

**Hallazgos:**

1. **Casi todas las filas son utilizables para pH y fecha.** Solo **11** de 92.738 filas tienen un pH que no se pudo convertir, y las fechas se interpretaron todas con `%d/%m/%Y`. Esto significa que la tabla puede usarse para análisis de pH y de tiempo casi completa, pero esas 11 filas no deben entrar a cálculos numéricos sin una decisión previa.

2. **Los 11 casos de pH sin número son textos con significado, no datos ausentes en origen.** Los ejemplos corresponden a valores censurados del laboratorio y a un error de digitación. Como hay **0** pH fuera de [0, 14], el total en revisión coincide con esos 11. La limitación es que los controles solo detectan lo que no se convierte o cae fuera del rango: un pH numérico plausible pero erróneo, dentro de [0, 14], no queda marcado.

## Entrega

El notebook completado se conserva como `soluciones/PD01_exploracion_agrosavia.ipynb`. La salida regenerable es `kit/salidas/soluciones/PD01/controles_agrosavia.csv`.

✅ Deben quedar visibles el perfil, los controles de conversión, la conciliación de filas y dos hallazgos. El notebook debe ejecutar todas las celdas en orden después de reiniciar el kernel.

⚠️ El intervalo de pH es un control de dominio para esta práctica. No determina por sí solo una recomendación agronómica. La frecuencia de análisis de suelo no representa automáticamente la distribución de los suelos de todo el territorio.